common/thorhelper/roxiehelper.cpp

```diff
@@ -411,8 +413,872 @@ bool CRHLimitedCompareHelper::getGroup(OwnedRowArray &group, const void *left)
     }
     return group.ordinality()>0;
 }
+
+//=========================================================================================
+
+// default implementations - can be overridden for efficiency...
+bool ISimpleInputBase::nextGroup(ConstPointerArray & group)
+{
+    // MORE - this should be replaced with a version that reads to a builder
+    const void * next;
+    while ((next = nextInGroup()) != NULL)
+        group.append(next);
+    if (group.ordinality())
+        return true;
+    return false;
+}
+
+void ISimpleInputBase::readAll(RtlLinkedDatasetBuilder &builder)
+{
+    loop
+    {
+        const void *nextrec = nextInGroup();
+        if (!nextrec)
+        {
+            nextrec = nextInGroup();
+            if (!nextrec)
+                break;
+            builder.appendEOG();
+        }
+        builder.appendOwn(nextrec);
+    }
+}
+
+
+//=========================================================================================
+
+// Ability to read an input stream and group and/or sort it on-the-fly
+
+using roxiemem::OwnedConstRoxieRow;
+
+class GroupedInputReader : public CInterfaceOf<IGroupedInput>
+{
+protected:
+    bool firstRead;
+    bool eof;
+    bool endGroupPending;
+    OwnedConstRoxieRow next;
+
+    IInputBase *input;
+    const ICompare *compare;
+public:
+    GroupedInputReader(IInputBase *_input, const ICompare *_compare)
+    : input(_input), compare(_compare)
+    {
+        firstRead = false;
+        eof = false;
+        endGroupPending = false;
+    }
+
+    virtual IOutputMetaData * queryOutputMeta() const
+    {
+        return input->queryOutputMeta();
+    }
+
+    virtual const void *nextInGroup()
+    {
+        if (!firstRead)
+        {
+            firstRead = true;
+            next.setown(input->nextInGroup());
+        }
+
+        if (eof || endGroupPending)
+        {
+            endGroupPending = false;
+            return NULL;
+        }
+
+        OwnedConstRoxieRow prev(next.getClear());
+        next.setown(input->nextUngrouped());  // skip incoming grouping if present
+
+        if (next)
+        {
+            assertex(prev);  // If this fails, you have an initial empty group. That is not legal.
+            if (compare && compare->docompare(prev, next) != 0) // MORE - could assert >=0, as input is supposed to be sorted
```

Comment:
Comment is not true in general - grouping doesn't nescessarily imply ordering (e.g., could be sorted in reverse or sorted with components in a different order).